src/main/cpp/windows/jssc.cpp

```diff
@@ -643,47 +644,81 @@ JNIEXPORT jobjectArray JNICALL Java_jssc_SerialNativeInterface_waitEvents
-            DWORD lpcbData;
-            DWORD result;
-            for(DWORD i = 0; i < keysCount; i++){
-                lpcchValueName = 256;
-                lpcbData = 256;
-                result = RegEnumValueA(phkResult, i, lpValueName, &lpcchValueName, NULL, NULL, lpData, &lpcbData);
-                if(result == ERROR_SUCCESS){
-                    env->SetObjectArrayElement(returnArray, i, env->NewStringUTF((char*)lpData));
+        else if(enumResult == ERROR_NO_MORE_ITEMS){
+            hasMoreElements = false;
+        }
+        else {
+            hasMoreElements = false;
+        }
+    }
+    if(keysCount > 0){
+        jclass stringClass = env->FindClass("java/lang/String");
+        returnArray = env->NewObjectArray((jsize)keysCount, stringClass, NULL);
+        DWORD lpcbData;
+        DWORD result;
+        /* iterate 2nd time but this time our array is ready to catch results. */
+        for(DWORD i = 0; i < keysCount; i++){
+            valueNameSize = 256;
+            lpcbData = lpData_capacity;
+            result = RegEnumValueA(phkResult, i, valueName, &valueNameSize, NULL, NULL, lpData, &lpcbData);
+            if(result == ERROR_SUCCESS){
+                env->SetObjectArrayElement(returnArray, i, env->NewStringUTF((char*)lpData));
+            }else if(result == ERROR_MORE_DATA && lpData_capacity < UINT_MAX){
+                /* whops our supplied buffer was not large enough. Fallback to a heap
+                 * allocd one. RegEnumValueA was kind enough to set 'lpcbData' to the
+                 * required length before return. */
+                lpData = (lpData == lpDataOnStack) ? NULL : lpData;
+                /* MS doc is unclear to me if returned size includes the required
+                 * zero-term. So to stay safe, we provide one byte more. */
+                if(lpcbData < UINT_MAX) lpcbData += 1;
+                byte *tmp = (byte*)realloc(lpData, lpcbData*sizeof*tmp);
+                if(tmp == NULL){
+                    jclass exClz = env->FindClass("java/lang/OutOfMemoryError");
+                    if(exClz != NULL) env->ThrowNew(exClz, NULL);
+                    returnArray = NULL;
+                    goto Finally;
                 }
+                /* install new buffer and reset 'i' so we try the same element again. */
+                lpData_capacity = lpcbData;
+                lpData = tmp;
+                i -= 1;
+            }else{
+                char exMsg[128];
+                snprintf(exMsg, sizeof exMsg, "RegEnumValueA(): Code %ld: "
+                    "https://learn.microsoft.com/en-us/windows/win32/debug/system-error-codes#system-error-codes",
+                    result);
+                jclass exClz = env->FindClass("java/lang/RuntimeException");
+                if(exClz != NULL) env->ThrowNew(exClz, exMsg);
+                returnArray = NULL;
+                goto Finally;
             }
         }
-        CloseHandle(phkResult);
     }
+Finally:
```

Comment:
Is it "normal" to use keyword "Finally" in gotos?  It seems like it may be confused with try/catch/finally, but I do not author enough C++ to know what is normal.

Reply:
My personal "normal" is naming the label "*endFn*" (aka end of this function). Because this matches the pattern I usually use. Where labels are named like: "*endFoo*", "*endSearchLoop*", ....

The "*Finally*" was a suggestion from another coder in another project. I don't know its origins.

Because of this I usually use "*endFn*" for my personal code and "*Finally*" for public/peer/shared code.

I did not yet hear any "*standard*" name for this label. One can find a decent amount of explanations about the *goto error handling* pattern in the web. But all those examples use some other label names. So my guess is that there is not really a "standard" for it :)